pkg/controllers/clusterresourceplacement/resource_selector.go

```diff
@@ -338,15 +354,55 @@ func generateManifest(object *unstructured.Unstructured) (*workv1alpha1.Manifest
 			}
 		}
 		if err != nil {
-			return nil, fmt.Errorf("failed to get the ports field in Serivce object, name =%s: %w", object.GetName(), err)
+			return nil, controller.NewUnexpectedBehaviorError(fmt.Errorf("failed to get the ports field in Serivce object, name =%s: %w", object.GetName(), err))
 		}
 	}
 
 	rawContent, err := object.MarshalJSON()
 	if err != nil {
-		return nil, fmt.Errorf("failed to marshal the unstructured object gvk = %s, name =%s: %w", object.GroupVersionKind(), object.GetName(), err)
+		return nil, controller.NewUnexpectedBehaviorError(fmt.Errorf("failed to marshal the unstructured object gvk = %s, name =%s: %w", object.GroupVersionKind(), object.GetName(), err))
+	}
+	return rawContent, nil
+}
+
+// generateManifest creates a manifest from the unstructured obj.
+func generateManifest(object *unstructured.Unstructured) (*workv1alpha1.Manifest, error) {
+	rawContent, err := generateRawContent(object)
+	if err != nil {
+		return nil, err
 	}
 	return &workv1alpha1.Manifest{
 		RawExtension: runtime.RawExtension{Raw: rawContent},
 	}, nil
 }
+
+// generateResourceContent creates a resource content from the unstructured obj.
+func generateResourceContent(object *unstructured.Unstructured) (*fleetv1beta1.ResourceContent, error) {
+	rawContent, err := generateRawContent(object)
+	if err != nil {
+		return nil, err
+	}
+	return &fleetv1beta1.ResourceContent{
+		RawExtension: runtime.RawExtension{Raw: rawContent},
+	}, nil
+}
+
+// selectResourcesForPlacement selects the resources according to the placement resourceSelectors.
+// It also generates an array of resource content based on the selected resources.
+func (r *Reconciler) selectResourcesForPlacement(placement *fleetv1beta1.ClusterResourcePlacement) ([]fleetv1beta1.ResourceContent, error) {
+	selectedObjects, err := r.gatherSelectedResource(placement.GetName(), placement.Spec.ResourceSelectors)
+	if err != nil {
+		return nil, err
+	}
+
+	resources := make([]fleetv1beta1.ResourceContent, len(selectedObjects))
+	for i, obj := range selectedObjects {
+		unstructuredObj := obj.DeepCopyObject().(*unstructured.Unstructured)
+		rc, err := generateResourceContent(unstructuredObj)
+		if err != nil {
+			return nil, err
```

Comment:
when will this encounter error? Should this be an "UnExpected" error?

Reply:
moved the unexpected error from internal func to here.